### PR TITLE
FedPM

### PR DESCRIPTION
**Pull request by yc7z**
# PR Type
[Feature]

# Short Description

This PR implements the "Federated Probabilistic Mask Training" algorithm described in http://arxiv.org/pdf/2209.15328. An example in the `examples/fedpm_example' folder runs this algorithm on the MNIST dataset, and it reproduces part of the experimental results in the original paper.

On a high level, the implementation consists mainly of the following components:
1. `masked_layers.py`: this module uses the linear and convolutional layers provided by PyTorch as a basis to implement masked versions of those layers, which ensures that only probability scores receive gradients.
2. `fedpm.py`: this module handles the server-side aggregation. The Bayesian aggregation approach described in the paper is implemented here.
3. The parameter exchanger part: this part is responsible for selecting the probability scores and then use them to sample binary masks to be sent to the server.

# Tests Added

Tests have been added to ensure that the newly added masked layers, parameter exchanger, strategy, and functions work correctly.

### CODE REVIEW

File: fl4health/model_bases/masked_layers.py
Comment (yc7z) [inline fl4health/model_bases/masked_layers.py:1]: I realize that there's a lot of "code duplication" here within the masked convolutional layers, but I couldn't think of a way to avoid this. An alternative is to implement a `MaskedConvNd" class and derive the specific masked conv layers from that, but I'm not sure that is actually better. Let me know what you think!

File: fl4health/parameter_exchange/layer_exchanger.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 from fl4health.parameter_exchange.partial_parameter_exchanger import PartialParameterExchanger
 
 TorchModule = TypeVar("TorchModule", bound=nn.Module)
-LayerSelectionFunction = Callable[[nn.Module, nn.Module], Tuple[NDArrays, List[str]]]
+LayerSelectionFunction = Callable[[nn.Module, Optional[nn.Module]], Tuple[NDArrays, List[str]]]
```
Comment (yc7z) [inline fl4health/parameter_exchange/layer_exchanger.py:12]: This change here is simply because I realized the old type annotation was not quite accurate.

File: fl4health/parameter_exchange/parameter_selection_criteria.py [hunk 2]
```diff
@@ -173,3 +176,43 @@ def smallest_increase_in_magnitude_scores(model: nn.Module, initial_model: Optio
...
+    parameters_to_exchange_names = []
+    parameters_to_exchange = []
+    model_states = model.state_dict()
+    with torch.no_grad():
+        if is_masked_module(model):
+            parameters_to_exchange_names.append("weight_scores")
+            weight_scores = model_states["weight_scores"]
+            # Sample a mask from the (unbounded) probability scores.
+            weight_probs = torch.sigmoid(weight_scores).cpu().numpy()
+            weight_mask_sample = bernoulli.rvs(weight_probs)
```
Comment (yc7z) [inline fl4health/parameter_exchange/parameter_selection_criteria.py:196]: Because of the Bernoulli sampling performed here, the parameters selected are in fact binary masks. That is, even though their corresponding names are something like "weight_scores" or "bias_scores", the tensors are not actually scores. But later after the tensors have been aggregated by the strategy, they will become score tensors again. I allowed this misalignment because these parameter names will later be needed to load the model anyway.

File: fl4health/utils/functions.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from typing import Any, Tuple
+
+import torch
+
+
+class BernoulliSample(torch.autograd.Function):
+    @staticmethod
+    def forward(bernoulli_probs: torch.Tensor) -> torch.Tensor:  # type: ignore
```
Comment (yc7z) [inline fl4health/utils/functions.py:8]: I ignored checking the type annotation because the parent class `torch.autograd.Function` is quite loose with its type annotations, while here we are certain that the input should be a tensor.

File: fl4health/parameter_exchange/fedpm_exchanger.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+class FedPmExchanger(DynamicLayerExchanger):
+    def pull_parameters(self, parameters: NDArrays, model: nn.Module, config: Optional[Config] = None) -> None:
+        current_state = model.state_dict()
+        layer_params, layer_names = self.unpack_parameters(parameters)
+        for layer_name, layer_param in zip(layer_names, layer_params):
+            # Apply the inverse of the Sigmoid function
+            # since the scores for masked layers are supposed to be unbounded.
+            with torch.no_grad():
+                current_state[layer_name] = sigmoid_inverse(torch.tensor(layer_param))
```
Comment (yc7z) [inline fl4health/parameter_exchange/fedpm_exchanger.py:19]: I also want to point out that it is actually important to apply this function to ensure that the parameter scores are unbounded (i.e., not bounded between 0 and 1), since during training there is no guarantee that the scores will remain between 0 and 1 after gradient updates.

File: examples/fedpm_example/server.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        config["downsampling_ratio"],
+        config["is_masked_model"],
+        config["priors_reset_frequency"],
+        local_epochs=config.get("local_epochs"),
+        local_steps=config.get("local_steps"),
+    )
+
+    initial_model = Masked4Cnn()
+
+    # Server performs simple FedAveraging as its server-side optimization strategy
```
Comment (emersodb) [inline examples/fedpm_example/server.py:54]: This comment is out of date I think. We can just move it to say FedPM here.

**Review by emersodb — COMMENTED** [review]
PR and Design look really awesome. Fantastic job on all the tests! I learned some new stuff about PyTorch too. Overall, everything looks pretty close to ready. Just some fairly minimal comments and discussion points.

**Comment by jewelltaylor** [comment]
Thanks for contributing Yuchong ❤️

File: examples/fenda_example/config.yaml [hunk 6]
```diff
@@ -3,7 +3,7 @@ n_server_rounds: 3 # The number of rounds to run FL
 
 # Parameters that describe clients
 n_clients: 3 # The number of clients in the FL experiment
-local_epochs: 1 # The number of epochs to complete for client
+local_epochs: 3 # The number of epochs to complete for client
```
[the remark this answers is not in the document]
Comment (yc7z) [reply examples/fenda_example/config.yaml:6]: I changed it to 3 because I wanted to produce similar experiment results in the paper, where the authors performed 3 local epochs.

File: fl4health/clients/fedpm_client.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+            metrics_reporter=metrics_reporter,
+        )
+
+    def setup_client(self, config: Config) -> None:
+        super().setup_client(config)
+        # Convert self.model to a masked model unless it is specified in the config
+        # file that the model is already a masked model.
+        if "is_masked_model" in config.keys():
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/clients/fedpm_client.py:42]: Yes I think it's fine to make this assumption too.

File: fl4health/strategies/fedpm.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+            on_fit_config_fn=on_fit_config_fn,
+            on_evaluate_config_fn=on_evaluate_config_fn,
+            accept_failures=accept_failures,
+            initial_parameters=initial_parameters,
+            fit_metrics_aggregation_fn=fit_metrics_aggregation_fn,
+            evaluate_metrics_aggregation_fn=evaluate_metrics_aggregation_fn,
+            weighted_aggregation=weighted_aggregation,
+            weighted_eval_losses=weighted_eval_losses,
+        )
+        self.weighted_aggregation = False
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/strategies/fedpm.py:87]: Yes good point!

File: fl4health/strategies/fedpm.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        names_to_layers: DefaultDict[str, List[NDArray]] = defaultdict(list)
+        total_num_clients: DefaultDict[str, int] = defaultdict(int)
+
+        # unpack the parameters and initialize the beta parameters to be all ones if they have not already
+        # been initialized.
+        for packed_layers, _ in results:
+            layers, names = self.parameter_packer.unpack_parameters(packed_layers)
+            for layer, name in zip(layers, names):
+                names_to_layers[name].append(layer)
+                total_num_clients[name] += 1
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/strategies/fedpm.py:135]: You are right that we could in theory omit the masks that are all zeros since they don't have any effect on aggregation, but the current implementation always communicates all the layers since there's no filtering being done in the parameter exchanger. I can't think of other scenarios where we might want to omit a layer. I can filter out the all-zero layers if you think that's better.

File: fl4health/strategies/fedpm.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+            self.beta_parameters[parameter_name] = (alpha_new, beta_new)
+            aggregation_result[parameter_name] = (alpha_new - 1) / (alpha_new + beta_new - 2)
+
+        return aggregation_result
+
+    def reset_beta_priors(self) -> None:
+        """
+        Reset the alpha and beta parameters for the Beta distribution to be arrays of all ones.
+        """
+        assert self.beta_parameters != {}
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/strategies/fedpm.py:160]: Ah yes you are right. I only added the assert to make sure we don't reset the priors when they haven't been initialized, but like you pointed out that's not necessary.

File: fl4health/parameter_exchange/parameter_selection_criteria.py [hunk 11]
```diff
@@ -173,3 +176,43 @@ def smallest_increase_in_magnitude_scores(model: nn.Module, initial_model: Optio
...
+            parameters_to_exchange.append(weight_mask_sample)
+            if "bias_scores" in model_states.keys():
+                parameters_to_exchange_names.append("bias_scores")
+                bias_scores = model_states["bias_scores"]
+                bias_probs = torch.sigmoid(bias_scores).cpu().numpy()
+                bias_mask_sample = bernoulli.rvs(bias_probs)
+                parameters_to_exchange.append(bias_mask_sample)
+        else:
+            for name, module in model.named_modules():
+                if is_masked_module(module):
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/parameter_exchange/parameter_selection_criteria.py:240]: Good point! I've rewritten these functions 🙂

File: fl4health/parameter_exchange/fedpm_exchanger.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+import torch
+import torch.nn as nn
+from flwr.common.typing import Config, NDArrays
+
+from fl4health.parameter_exchange.layer_exchanger import DynamicLayerExchanger
+from fl4health.utils.functions import sigmoid_inverse
+
+
+class FedPmExchanger(DynamicLayerExchanger):
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/parameter_exchange/fedpm_exchanger.py:12]: Thanks for bringing up the alternative! After thinking about it I decided to go with the selection function approach because I think it extends the existing functionality quite nicely. However, I rewrote the `__init__` method for `FedPMExchanger` since it is designed to be used specifically with `select_scores_and_sample_masks`, while originally the user needs to pass in the selection function themselves. I think now it makes more sense and hopefully reduces some confusion 🙂

File: poetry.lock
[the remark this answers is not in the document]
Comment (yc7z) [reply poetry.lock:1]: Yeah that's strange because I don't think I had to bump any library. I'll try using the current lock file.

File: examples/fenda_example/config.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (emersodb) [reply examples/fenda_example/config.yaml:6]: The reason I ask is that this is the config for the FENDA example?

File: fl4health/strategies/fedpm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/strategies/fedpm.py:136]: We can leave the current implementation. Maybe just leave a note in the comments that we don't presently exchange a subset of layers but might in the future?

File: fl4health/parameter_exchange/fedpm_exchanger.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/parameter_exchange/fedpm_exchanger.py:12]: Sounds good!

File: fl4health/model_bases/masked_layers.py [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+
+    Attributes:
+        weight: weights of the module.
+        bias:  bias of the module.
+        weight_score: learnable scores for the weights. Has the same shape as weight.
+        bias_score: learnable scores for the bias. Has the same shape as bias.
+    """
+
+    __constants__ = ["in_features", "out_features"]
+    in_features: int
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/model_bases/masked_layers.py:41]: They were declared in `nn.Linear` in this way so I just followed it without thinking too much. But after looking into it this is what I've found:

1. `__constants__` is an annotation from [TorchScript](https://pytorch.org/docs/stable/jit.html#constants) that marks certain members of a class as fixed. This can inform the PyTorch JIT (Just-In-Time) compiler to make certain assumptions for optimization purposes.
2. I couldn't find out the reason for declaring `weight: Tensor` before `__init__`. Normally the fields declared before `__init__` are static attributes belonging to the class rather than instance attributes, but here we declare `weight` as an instance attribute afterwards anyway, so maybe it was done for type annotation and better readability?

In any case I'm removing these declarations because they have already been done in the parent class

File: fl4health/model_bases/masked_layers.py [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+        out_features: int,
+        bias: bool = True,
+        device: Optional[torch.device] = None,
+        dtype: Optional[torch.dtype] = None,
+    ) -> None:
+        super().__init__(in_features, out_features, bias, device, dtype)
+        self.in_features = in_features
+        self.out_features = out_features
+        self.weight = Parameter(
+            torch.empty((out_features, in_features), device=device, dtype=dtype), requires_grad=False
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/model_bases/masked_layers.py:57]: No you are right, this was just an oversight on my part. Using `torch.empty` here is not necessary and would've overridden what happens in `super().__init__()`. I think originally I wanted to modify `self.weight` so that it doesn't receive gradient, but I could've simply reset `weight.requires_grad` instead (this is also how I handled the convolutional layers).

File: fl4health/model_bases/masked_layers.py [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+            masked_bias = bias_mask * self.bias
+        else:
+            masked_bias = None
+        return self._conv_forward(input, weight=masked_weight, bias=masked_bias)
+
+    @classmethod
+    def from_pretrained(cls, conv_module: nn.Conv1d) -> "MaskedConv1d":
+        has_bias = conv_module.bias is not None
+        # we create new variables below to make mypy happy since kernel_size has
+        # type Union[int, Tuple[int]] and kernel_size_ has type Tuple[int]
```
[the remark this answers is not in the document]
Comment (yc7z) [reply fl4health/model_bases/masked_layers.py:158]: 😂

File: examples/fenda_example/config.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (yc7z) [reply examples/fenda_example/config.yaml:6]: I don't know why I didn't notice this 😂 I've changed it back.

File: fl4health/model_bases/masked_layers.py [hunk 16]
```diff
@@ -50,12 +21,35 @@ def __init__(
...
         super().__init__(in_features, out_features, bias, device, dtype)
         self.in_features = in_features
         self.out_features = out_features
-        self.weight = Parameter(
-            torch.empty((out_features, in_features), device=device, dtype=dtype), requires_grad=False
-        )
+        self.weight.requires_grad = False
         self.weight_scores = Parameter(torch.randn_like(self.weight), requires_grad=True)
         if bias:
             self.bias = Parameter(torch.empty(out_features, device=device, dtype=dtype), requires_grad=False)
```
Comment (emersodb) [inline fl4health/model_bases/masked_layers.py:55]: Since we're dropping the empyy for weight and just setting `weight.requires_grad = False`, we should probably do that here too yes?

File: fl4health/model_bases/masked_layers.py [hunk 17]
```diff
@@ -99,7 +96,40 @@ def from_pretrained(cls, linear_module: nn.Linear) -> "MaskedLinear":
 
 class MaskedConv1d(nn.Conv1d):
     """
-    Implementation of masked Conv1d layers in a manner that is analogous to MaskedLinear.
+    Implementation of masked Conv1d layers.
```
Comment (emersodb) [inline fl4health/model_bases/masked_layers.py:99]: We've been putting class documentation in the init files. I'm not sure if that's a good practice or not, but it's where we've been doing it. Do you think we should start doing "class" level docs instead?

File: fl4health/model_bases/masked_layers.py (hunk 16 above)
Comment (yc7z) [reply fl4health/model_bases/masked_layers.py]: Yes you are right.

File: fl4health/model_bases/masked_layers.py (hunk 17 above)
Comment (yc7z) [reply fl4health/model_bases/masked_layers.py]: I've seen both practices and I'm not sure if one way is more standard than the other. I'm moving this to `__init__` for consistency for now.

File: fl4health/parameter_exchange/parameter_selection_criteria.py [hunk 18]
```diff
@@ -178,41 +177,72 @@ def smallest_increase_in_magnitude_scores(model: nn.Module, initial_model: Optio
...
+
+def _process_masked_module(
+    module: nn.Module, model_state_dict: Dict[str, Tensor], module_name: Optional[str] = None
+) -> Tuple[NDArrays, List[str]]:
+    """
+    Perform Bernoulli sampling using the weight and bias scores of a masked module.
+
+    Args:
+        module (nn.Module): the module upon which operations described above are performed.
+            "module" can either be a submodule of the model trained in FedPM, or it can be the model itself.
```
Comment (emersodb) [inline fl4health/parameter_exchange/parameter_selection_criteria.py:196]: I might suggest a small change to this comment. Rather than saying "it can be the model itself" I would suggest, "it can a standalone module itself" I got a little confused and thought that you intended this to work for a full model being passed in as `module`, and couldn't figure out how that would work. However, in visiting the other code, I saw what you meant.

File: fl4health/model_bases/masked_layers.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/model_bases/masked_layers.py]: Sounds good

**Comment by emersodb** [comment]
I think the only minor question left before this is read to go is whether the lock file really needs to change (since no toml updates have happened). Outside of that, this is ready to go!

File: poetry.lock
[the remark this answers is not in the document]
Comment (yc7z) [reply poetry.lock:1]: Everything seems fine with the current lock file, so I think we're good sticking with it.

**Review by emersodb — APPROVED** [review]
Ready to go!
